File: 2021/10/11/MiddleNode.java
/**
 * https://leetcode.com/problems/middle-of-the-linked-list/
 * @author Andrew Jarombek
 * @since 10/11/2021
 */

public class MiddleNode {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() { }
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode middleNode(ListNode head) {
        ListNode mid = head;
        head = head.next;

        int index = 1;

        while (head != null) {
            index++;

            if (index % 2 == 0) {
                mid = mid.next;
            }

            head = head.next;
        }

        return mid;
    }
}

File: 2022/05/10/middleNode.ts
/**
 * https://leetcode.com/problems/middle-of-the-linked-list/
 * @author Andrew Jarombek
 * @since 5/10/2022
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function middleNode(head: ListNode | null): ListNode | null {
    let slow = head;
    let fast = head;

    while (fast?.next != null) {
        slow = slow.next;
        fast = fast.next?.next;
    }

    return slow;
}